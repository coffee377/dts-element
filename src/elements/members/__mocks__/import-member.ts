import {Member} from '../../member';

export class ImportMember extends Member<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[ImportMember ${this.parameters.owned.parameters.name}]`;
  }

}
