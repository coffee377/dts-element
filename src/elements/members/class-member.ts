import {Container} from '../../collections';
import {emit_abstract} from '../../helpers/emit-abstract';
import {emit_accessor} from '../../helpers/emit-accessor';
import {emit_readonly} from '../../helpers/emit-readonly';
import {emit_static} from '../../helpers/emit-static';
import {Member} from '../member';

export type ClassAccessor = 'private' | 'protected' | 'public' | null;

// tslint:disable-next-line no-empty-interface
export interface IClassMemberRequiredParameters {}

export interface IClassMemberOptionalParameters {
  accessor: ClassAccessor;
  abstract: boolean;
  readonly: boolean;
  static: boolean;
}

export class ClassMember extends Member<IClassMemberRequiredParameters, IClassMemberOptionalParameters> {

  public get default_parameters(): IClassMemberOptionalParameters {
    return {
      accessor: null,
      abstract: false,
      readonly: false,
      static: false,
    };
  }

  public _emit(_container: Container): string {
    const {owned} = this.parameters;
    const accessor = emit_accessor(this.parameters.accessor, owned);
    const readonly = emit_readonly(this.parameters.readonly, owned);
    const abstract = emit_abstract(this.parameters.abstract, owned);
    const a_static = emit_static(this.parameters.static, owned);
    return `${owned._emit_jsdoc()}${accessor}${a_static}${abstract}${readonly}${owned._emit_raw(this)}`;
  }

}
