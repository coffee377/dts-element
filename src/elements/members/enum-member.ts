import {emit_declaration_name} from '../../helpers/emit-declaration-name';
import {emit_equal} from '../../helpers/emit-equal';
import {Stack} from '../../stack';
import {VariableDeclaration} from '../declarations/variable-declaration';
import {Member} from '../member';

export interface IEnumMemberRequiredParameters {
  owned: VariableDeclaration;
}

export interface IEnumMemberOptionalParameters {
  value: number | null;
}

export class EnumMember extends Member<IEnumMemberRequiredParameters, IEnumMemberOptionalParameters> {

  private _instance_of_enum_member: true;

  public get default_parameters(): IEnumMemberOptionalParameters {
    return {
      value: null,
    };
  }

  public _emit(stack: Stack, fn: (value: number | null) => number): string {
    const {owned} = this.parameters;
    const value = fn(this.parameters.value);
    return `${owned._emit_jsdoc()}${emit_declaration_name(owned.parameters.name, stack)}${emit_equal(value, stack)}`;
  }

}
