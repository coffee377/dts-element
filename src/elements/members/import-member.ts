import {emit_name_as} from '../../helpers/emit-name-as';
import {Stack} from '../../stack';
import {Declaration} from '../declaration';
import {Member} from '../member';

export interface IImportMemberRequiredParameters {
  owned: Declaration;
}

export interface IImportMemberOptionalParameters {
  name: string | null;
}

export class ImportMember
    extends Member<IImportMemberRequiredParameters, IImportMemberOptionalParameters> {

  public get default_parameters(): IImportMemberOptionalParameters {
    return {
      name: null,
    };
  }

  public _emit(_stack: Stack): string {
    const {owned} = this.parameters;
    return `${emit_name_as(this.parameters.name)}${owned.parameters.name}`;
  }

}
