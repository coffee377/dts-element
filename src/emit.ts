import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
