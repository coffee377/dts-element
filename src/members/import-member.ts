import * as ts from 'typescript';
import {ElementKind} from '../constants';
import {create_element, IElement, IElementOptions} from '../element';
import {create_identifier_if_defined} from '../utils';

export interface IImportMemberOptions extends IElementOptions {
  name: string;
  property?: string;
}

export interface IImportMember
  extends IElement<ElementKind.ImportMember>, IImportMemberOptions {}

export const create_import_member = (options: IImportMemberOptions): IImportMember => ({
  ...create_element(ElementKind.ImportMember),
  ...options,
});

export const transform_import_member = (element: IImportMember, path: IElement<any>[]) =>
  ts.createImportSpecifier(
    /* propertyName  */ create_identifier_if_defined(element.property),
    /* name          */ ts.createIdentifier(element.name),
  );
