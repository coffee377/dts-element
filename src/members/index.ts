export {IClassMember, IClassMemberOptions, create_class_member} from './class-member';
export {IImportMember, IImportMemberOptions, create_import_member} from './import-member';
export {IObjectMember, IObjectMemberOptions, create_object_member} from './object-member';
