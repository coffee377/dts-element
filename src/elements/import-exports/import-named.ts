import {emit_elements} from '../../helpers/emit-elements';
import {Stack} from '../../stack';
import {ImportExport} from '../import-export';
import {ImportMember} from '../members/import-member';

export interface IImportNamedRequiredParameters {
  from: string;
}

export interface IImportNamedOptionalParameters {
  members: ImportMember[];
}

/**
 * ```ts
 * import { some_declaration } from "path/to/some";
 * ```
 */
export class ImportNamed extends ImportExport<IImportNamedRequiredParameters, IImportNamedOptionalParameters> {

  private _instance_of_import_named: true;

  public get default_parameters(): IImportNamedOptionalParameters {
    return {
      members: [],
    };
  }

  public _emit(stack: Stack): string {
    const {members, from} = this.parameters;
    return `import {${emit_elements(members, stack, ', ')}} from ${JSON.stringify(from)};`;
  }

}
