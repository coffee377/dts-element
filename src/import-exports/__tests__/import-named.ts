import {emit} from '../../emit';
import {create_import_member} from '../../members/import-member';
import {create_import_named} from '../import-named';

it('should return correctly with from', () => {
  expect(emit(
    create_import_named({
      from: 'path/to/somewhere',
    }),
  )).toMatchSnapshot();
});

it('should return correctly with from, name', () => {
  expect(emit(
    create_import_named({
      from: 'path/to/somewhere',
      name: 'abc',
    }),
  )).toMatchSnapshot();
});

it('should return correctly with from, members', () => {
  expect(emit(
    create_import_named({
      from: 'path/to/somewhere',
      members: [
        create_import_member({name: 'a'}),
        create_import_member({name: 'b'}),
      ],
    }),
  )).toMatchSnapshot();
});

it('should return correctly with from, name, members', () => {
  expect(emit(
    create_import_named({
      from: 'path/to/somewhere',
      name: 'abc',
      members: [
        create_import_member({name: 'a'}),
        create_import_member({name: 'b'}),
      ],
    }),
  )).toMatchSnapshot();
});
