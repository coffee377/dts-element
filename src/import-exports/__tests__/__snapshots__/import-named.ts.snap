// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with from 1`] = `"import \\"path/to/somewhere\\";"`;

exports[`should return correctly with from, members 1`] = `"import { a, b } from \\"path/to/somewhere\\";"`;

exports[`should return correctly with from, name 1`] = `"import abc from \\"path/to/somewhere\\";"`;

exports[`should return correctly with from, name, members 1`] = `"import abc, { a, b } from \\"path/to/somewhere\\";"`;
